sklearn/preprocessing/data.py

```diff
@@ -1543,8 +1543,45 @@ def transform(self, X):
             else:
                 XP = np.empty((n_samples, self.n_output_features_),
                               dtype=X.dtype, order=self.order)
-                for i, comb in enumerate(combinations):
-                    XP[:, i] = X[:, comb].prod(1)
+
+                # What follows is a faster implementation of:
```

Comment:
At the moment I don't find the algorithm easy to read. I think if you explicitly say "dynamic programming" it might be clearer that your algorithm is building subsequent columns from precomputed partial solutions.